File: src/com/alchitry/labs/Locations.java
package com.alchitry.labs;

import java.io.File;
import java.net.URISyntaxException;

import com.alchitry.labs.gui.main.MainWindow;

public class Locations {
	public static final File progDir;
	public static final String progPrefix;

	static {
		File prog = null;
		try {
			prog = new File(MainWindow.class.getProtectionDomain().getCodeSource().getLocation().toURI());

		} catch (URISyntaxException e1) {
			e1.printStackTrace();
			Util.log.severe("Could not detect program directory!");
		}

		prog = prog.getParentFile();
		if (prog != null && !new File(prog.getPath() + File.separator + "lib").exists())
			progDir = prog.getParentFile();
		else
			progDir = prog;

		if (progDir != null)
			progPrefix = progDir.getPath() + File.separatorChar;
		else
			progPrefix = "";

	}

	public static final File LIBRARY = new File(progPrefix + "library");
	public static final File BASE = new File(LIBRARY + File.separator + "base");
	public static final File COMPONENTS = new File(LIBRARY + File.separator + "components");
	public static final File FIRMWARE = new File(LIBRARY + File.separator + "firmware");

	public static final File TOOLS = new File(progPrefix + "tools");
	public static final File BIN = new File(TOOLS + File.separator + Util.osDir + File.separator + "bin");
	public static final File ETC = new File(TOOLS + File.separator + "etc");
	public static final File LIB = new File(TOOLS + File.separator + Util.osDir + File.separator + "lib");

	public static final File RESOURCES = new File(progPrefix + "res");

	public final static File TEMPLATE_DIR = new File(BASE + File.separator + "templates");

	private Locations() {
	}
}

File: src/com/alchitry/labs/Locations.kt
package com.alchitry.labs

import com.alchitry.labs.Util.logger
import com.alchitry.labs.Util.osDir
import com.alchitry.labs.gui.main.MainWindow
import java.io.File
import java.net.URISyntaxException
import java.net.URL

object Locations {
    @JvmField
    val progDir: File?
    @JvmField
    val progPrefix: String
    @JvmField
    val LIBRARY: File
    @JvmField
    val BASE: File
    @JvmField
    val COMPONENTS: File
    @JvmField
    val FIRMWARE: File
    @JvmField
    val TOOLS: File
    @JvmField
    val BIN: File
    @JvmField
    val ETC: File
    @JvmField
    val LIB: File
    @JvmField
    val RESOURCES: File
    @JvmField
    val TEMPLATE_DIR: File


    init {
        // env LUCIDHDL_HOME or current folder to find library/components
        val home = System.getenv("LUCIDHDL_HOME")
        progDir = File(home ?: System.getProperty("user.dir"))
        progPrefix = progDir?.path ?: ""

        LIBRARY = Util.assembleFile(progPrefix, "library")
        BASE = Util.assembleFile(LIBRARY, "base")
        COMPONENTS = Util.assembleFile(LIBRARY, "components")
        FIRMWARE = Util.assembleFile(LIBRARY, "firmware")
        TOOLS = Util.assembleFile(progPrefix + "tools")
        BIN = Util.assembleFile(TOOLS, (osDir ?: ""), "bin")
        ETC = Util.assembleFile(TOOLS, "etc")
        LIB = Util.assembleFile(TOOLS, (osDir ?: ""), "lib")
        RESOURCES = Util.assembleFile(progPrefix + "res")
        TEMPLATE_DIR = Util.assembleFile(BASE, "templates")
    }
}